### PR TITLE
remove #ifdefs from HAMOCC

### PR DESCRIPTION
**Pull request by mvertens**
This PR is a substitute for #264 (which I would like to close). It does not build upon the ability for CMEPS to calculate DMS and BROMO fluxes. 
The current testing I have done is to compare relative to the latest baseline:
1. ./create_test  --xml-category aux_blom_noresm --xml-machine betzy --xml-compiler intel --gen aux_blom_noresm_13102023 --baseline-root /cluster/shared/noresm/noresm_baselines --compare aux_blom_noresm_27092023 --project nn2345k
The following tests were run:
  ERP_Ld3.T62_tn14.NOINY.betzy_intel  - passed
  ERS_Ld3.T62_tn14.NOINYOC.betzy_intel - passed
  ERS_Ld3.T62_tn14_wtn14nw.NOINY_WW3.betzy_intel - failed baseline comparison
  ERS_Ld3.T62_tn14_wtn14nw.NOINY_WW3.betzy_intel.blom-wavice - failed baseline comparison
  SMS_D_Ld1.T62_tn14.NOINYOC.betzy_intel - passed
I think there is something problematic with the current ww3dev code that is leading to failures in the baseline comparison. The same tests without ww3 passed with no problem. I will raise an issue on this - but I don't think this should hold up this PR.
2. Ran two tests using noresm2.0.6 - SMS.T62_tn14.NOINYOC.betzy_intel. The first used used BLOM master and the second tests used this PR for BLOM. The results were bfb after 5 days.
3. Ran two fully coupled tests using noresm2.0.6 - SMS.f19_tn14.N1850esm.betzy_intel. The first used used BLOM master and the second tests used this PR for BLOM. The results were bfb after 5 days.

### CODE REVIEW

**Comment by mvertens** [comment]
@JorgSchwinger - I ran a fully coupled test in the noresm2.0.6 code base - as was suggested-  and the results were bfb.

File: cime_config/buildnml.csh
Comment (JorgSchwinger) [inline cime_config/buildnml.csh:1]: This file was removed previously, I believe it shouldn't be re-introduced through this PR?

File: hamocc/aufr_bgc.F90
Comment (JorgSchwinger) [inline hamocc/aufr_bgc.F90]: This comment shouldn't be re-introduced, the burial fields for C-isotopes are now in place (`bur_*` below)

File: hamocc/aufr_bgc.F90
Comment (JorgSchwinger) [inline hamocc/aufr_bgc.F90:612]: This loop is now missing!!! Needs to be added again

File: hamocc/aufr_bgc.F90
Comment (JorgSchwinger) [inline hamocc/aufr_bgc.F90:589]: Loop for burial variables should be added here (see comment above)!

File: hamocc/aufw_bgc.F90
Comment (JorgSchwinger) [inline hamocc/aufw_bgc.F90]: This should be correct (the trc array carries 2 time-levels of tracer values in the 3rd dimension, these are both copied into locetra here). 

Not sure though - wasn't it necessary to have locetra as allocatable array? If it works it's fine

File: hamocc/carchm.F90
Comment (JorgSchwinger) [inline hamocc/carchm.F90]: ks isn't a local variable, it needs to be imported from `mo_param1_bgc`

File: hamocc/hamocc_init.F90
Comment (JorgSchwinger) [inline hamocc/hamocc_init.F90]: `init_indices` doesn't seem to be used?

File: cime_config/buildnml.csh
Comment (mvertens) [reply cime_config/buildnml.csh:1]: Done.

File: hamocc/aufr_bgc.F90
Comment (mvertens) [reply hamocc/aufr_bgc.F90]: Done.

File: hamocc/aufr_bgc.F90
Comment (mvertens) [reply hamocc/aufr_bgc.F90:589]: Done. Thanks for catching this. It was not picked up in my git merge.

File: hamocc/aufw_bgc.F90
Comment (mvertens) [reply hamocc/aufw_bgc.F90]: @JorgSchwinger - locetra works now as a non-allocatable array on the stack.

File: hamocc/carchm.F90
Comment (mvertens) [reply hamocc/carchm.F90]: Thanks for catching this. I'm concerned that the testing we have is not really covering all the use cases that we need to have identified these problems. It would be good to add a few tests that would create baselines and would have caught these problems as part of the regression testing. A quick chat to out outline what we want to do moving forwards would be very helpful. I'm happy to implement that.

File: hamocc/mo_control_bgc.F90
Comment (JorgSchwinger) [inline hamocc/mo_control_bgc.F90:49]: I think it is actually correct to have save attributes here. Unless a module variable is used in the main program it can go out of scope and the Fortran standard doesn't require that the variable's value is retained (as far as I know).

Comment (mvertens) [reply hamocc/mo_control_bgc.F90:49]: @JorgSchwinger - for module scope my understanding is that save is now on by default.

Comment (gold2718) [reply hamocc/mo_control_bgc.F90:49]: > @JorgSchwinger - for module scope my understanding is that save is now on by default.

This is true. Note that this is a change from the behavior in Fortran 2003 and before. Beginning with Fortran 2008 (and universally true in contemporary compilers to my knowledge), all module-scope variables have the save attribute by default.

File: trc/mod_tracers.F90
Comment (JorgSchwinger) [inline trc/mod_tracers.F90:103]: I would find it preferable if the `init_indices` routine would be called in `hamocc_init`, but I guess this isn't possible because this routine is called first? But we should at least have a comment in `hamocc_init` saying that the initialization of HAMOCC indices is done here.

File: phy/mod_ifdefs.F90
Comment (JorgSchwinger) [inline phy/mod_ifdefs.F90:1]: I this file needed? BLOM ifdefs aren't touched in this PR?

File: hamocc/mo_control_bgc.F90
Comment (mvertens) [reply hamocc/mo_control_bgc.F90:49]: @JorgSchwinger - CESM has been leverage this to remove save attributes from module variables in new development versions of the code.

**Review by JorgSchwinger — APPROVED** [review]
Thanks a lot for this effort! It's great to get this into HAMOCC! I agree that at least for this very invasive PR a testing with more (or all) of the options activated would be good. I'm happy to have a chat on this.

Maybe also good if @jmaerz or @TomasTorsvik could have a second review. It's a lot of stuff and I might have missed something.

File: hamocc/mo_control_bgc.F90
Comment (JorgSchwinger) [reply hamocc/mo_control_bgc.F90:49]: Ok, thanks for clarifying this - I see my Fortran knowledge is becoming outdated...

File: phy/mod_ifdefs.F90
Comment (mvertens) [reply phy/mod_ifdefs.F90:1]: You still need this file - since these are the #ifdefs outside of HAMOCC that I have not touched yet.
I would be happy to tackle this - but a lot of the files where these appear are not .F90 files and also not modules.
I'd like to see that happen - but it seemed too invasive for this PR.

File: trc/mod_tracers.F90
Comment (mvertens) [reply trc/mod_tracers.F90:103]: Good point. I'll add that statement there.

File: hamocc/mo_apply_rivin.F90
Comment (jmaerz) [inline hamocc/mo_apply_rivin.F90:97]: minor, but could be added to l. 89 above

File: .github/workflows/ci.yml
Comment (TomasTorsvik) [inline .github/workflows/ci.yml:1]: This file is removed in commit 5957d2c25a31705778a6fb7dae06e8d0513139e4
What was the background for this? If it is not working, is there some way to disable CI functionality instead of removing it completely?

File: trc/restart_trcrd.F90
Comment (TomasTorsvik) [inline trc/restart_trcrd.F90:40]: Would it make sense to move this to `mod_ifdefs.F90`, and make use of `use_hamocc`in both `restart_trcrd.F90` and `restart_wt.F`? Not sure if this is a good idea, if the long term plan is to get rid of `mod_ifdefs.F90`.

File: hamocc/carchm.F90
Comment (jmaerz) [inline hamocc/carchm.F90:25]: I would suggest to keep the `pflxdms` and `pflxbromo` out of this PR, since they will be included via the nuopc PR, right? At the moment, they aren't used.

File: hamocc/inventory_bgc.F90
Comment (jmaerz) [inline hamocc/inventory_bgc.F90:364]: Here and also for the other species: Instead of repeating all contributions, I would suggest to write once the general contributions and then, afterward:
```
if (use_boxatm) then
        totalcarbon = totalcarbon + zatmco2*ppm2con
else
        totalcarbon = totalcarbon + co2flux
endif
```
which is easier to maintain, shorter and clearer, what the difference is.

File: hamocc/mo_param1_bgc.F90
Comment (jmaerz) [inline hamocc/mo_param1_bgc.F90:50]: Two things: I would suggest to give at least one comment per group (e.g.: `! Base tracer` etc. Then, is there any reason not to at least add a `protect` statement for the indices values?

File: hamocc/inventory_bgc.F90
Comment (jmaerz) [inline hamocc/inventory_bgc.F90]: I am a bit surprised to see the `-1` here. Is it really needed? - any tracers to be calculated should be `>0` (and loops also start at 1, where these values are calculated)

File: hamocc/mo_param_bgc.F90
Comment (jmaerz) [inline hamocc/mo_param_bgc.F90]: I am running into compilation problems - the line with dustd1, dustd2, dustsink needs to be deleted. (They are introduced now via `mo_biomod` - it was formerly a bit twisted written.)

File: hamocc/carchm.F90
Comment (mvertens) [reply hamocc/carchm.F90:25]: Good point. I've removed them.

**Comment by jmaerz** [comment]
Out of curiosity: what was/is the purpose of `ci.yml` that has been removed - is this the setup to do the testing on github? - if so, why was it now removed?

File: hamocc/inventory_bgc.F90
Comment (mvertens) [reply hamocc/inventory_bgc.F90:364]: Great idea! I'll implement that.

**Review by TomasTorsvik — APPROVED** [review]

**Comment by jmaerz** [comment]
@mvertens , @JorgSchwinger , just to let you know, with #282, we updated the `beyond-CMIP6` branch with the `master` - so that this PR can go next into `master`.

File: hamocc/mo_param_bgc.F90
Comment (mvertens) [reply hamocc/mo_param_bgc.F90]: I don't understand this comment. I have no problems compiling.

Comment (jmaerz) [reply hamocc/mo_param_bgc.F90]: Hi @mvertens , `dustd1, dustd2, dustsink` are already included by `use mo_biomod, only:` above so that at least on my side, the compiler complains `Symbol ‘dustd1’ at (1) conflicts with symbol from module ‘mo_biomod’, use-associated at (2)` (compiled locally via meson). Similarly, `dustd2` and `dustsink` are also already defined. As written, it was very twisted before (these were only included via `use` when `AGG` was switched on, while then the local variables were required inside the routine for some initialization). Deletion of the line 
```
  REAL :: dustd1, dustd2, dustsink !AGG
```
should do the job.

**Comment by mvertens** [comment]
As an additional test - I compare master versus this PR for a SMS_D_Ld1.T62_tn14.NOINYOC.betzy_intel - but where I had the following xml variables set to TRUE:
	HAMOCC_CFC: TRUE
	HAMOCC_CISO: TRUE
	HAMOCC_NATTRC: TRUE
	HAMOCC_SEDBYPASS: TRUE
And the results were bit-for-bit in comparing the blom history files
       *.hd.0001-01-01.nc
        *.hbgcd.0001-01-01.nc
So this gives me more confidence that the tests that @JorgSchwinger was proposing will pass.

**Comment by mvertens** [comment]
Once I reintroduce ci.yml - the test is now failing. In the past I did notice that periodically it would fail and then pass.

File: trc/restart_trcrd.F90
Comment (mvertens) [reply trc/restart_trcrd.F90:40]: Could we defer deciding this to #281? I'd like to know from @matsbn with always compiling the hamocc code since there are no longer any #ifdefs.

File: hamocc/mo_param_bgc.F90
Comment (mvertens) [reply hamocc/mo_param_bgc.F90]: Yes - you are right. Sorry for not seeing that.

Comment (jmaerz) [reply hamocc/mo_param_bgc.F90]: No worries - I am happy to be able to help :-)

Comment (mvertens) [reply hamocc/mo_param_bgc.F90]: I've pushed the fix back.

**Review by jmaerz — APPROVED** [review]

File: trc/restart_trcrd.F90
Comment (TomasTorsvik) [inline trc/restart_trcrd.F90:95]:
```suggestion
#ifdef HAMOCC
     call hamocc_init(1,rstfnm_hamocc)
#endif
```

The CI intel compiler build fails on the call to hamocc_init. When meson builds without the 'ecosys' option it does not compile code in the `hamocc` directory, so then `hamocc_init` is not available. 

It may be safer for now to keep the calls to hamocc funtions/subroutines in #ifdef, when called from outside hamocc.

Comment (TomasTorsvik) [reply trc/restart_trcrd.F90:95]: I tried to update the ci.yml file to include tests with 'ecosys' enabled, but the tests are still failing. Not sure what's going on, but the meson build system doesn't seem to recognize `ecosys=true`.

Comment (mvertens) [reply trc/restart_trcrd.F90:95]: @TomasTorsvik - thanks for looking into this. I agree that it makes sense at this point to keep the #ifdef HAMOCC on.
Are we okay with merging this PR - or do we want to resolve the meson build system problem first?

Comment (TomasTorsvik) [reply trc/restart_trcrd.F90:95]: We can merge this now. I can add the meson build afterwards, I didn't enable it properly with ecosys==true.

**Comment by JorgSchwinger** [comment]
@mvertens, @jmaerz, @TomasTorsvik I just tested this PR, and the results for the C-isotopes are NOT bfb!

I'm investigating this further, but this should be clarified before we move on

**Comment by mvertens** [comment]
@JorgSchwinger - thanks for doing this test. Can you please point me to your rundir and caseroot.
I thought I had turned on isotopes last night and it was bfb - but clearly I'm not doing something right.

**Comment by TomasTorsvik** [comment]
@mvertens , @JorgSchwinger , @jmaerz 
I will open a new issue on this, with reference to PR #280. We will probably need to do a fix in a new PR anyway, so it's not convenient to keep the discussion in an already merged PR.

**Pull request merged**
